File: Graphs/Graph.js
"use strict";
const LinkedList = require('./LinkedList.js');
const Node = require('./Node.js');

module.exports = class Graph {
    constructor(vertices) {
        this.vertices = vertices;
        this.list = [];
        
        var it;
        for (it = 0; it < vertices; it++) {
            let temp = new LinkedList(); // Initializes a LinkedList with head -> null
            this.list.push(temp);
        }
    }

    addEdge(source, destination) {
        if (source < this.vertices && destination < this.vertices) {
            this.list[source].insertAtHead(destination);
        }
        return this;
    }

    printGraph() {
        console.log(">>Adjacency List of Directed Graph<<");
        var i;
        for (i = 0; i < this.list.length; i++) {
            console.log("|" + String(i) + "| => ");
            let temp = this.list[i].getHead();
            while (temp != null) {
                console.log("[" + String(temp.data) + "] -> ");
                temp = temp.nextElement;
            }
            console.log("null ");
        }
    }

    strGraph() {
        let str = "";
        var i;
        for (i = 0; i < this.list.length; i++) {
            str = str + "|" + String(i) + "| => ";
            let temp = this.list[i].getHead();
            while (temp != null) {
                str += ("[" + String(temp.data) + "] -> ");
                temp = temp.nextElement;
            }
            str += "null ";
        }
        return str;
    }
}

File: Graphs/Graph.java
package Graphs;

public class Graph {

    int vertices; // Total number of vertices in graph
    DoublyLinkedList<Integer> adjacencyList[]; // An array of linked lists to store adjacent vertices.
    
    @SuppressWarnings("unchecked")
    public Graph(int vertices) {
        this.vertices = vertices;
        adjacencyList = new DoublyLinkedList[vertices];
        // DoublyLinkedList instance for each vertex will be store in this array adjacencyList
        for (int i = 0; i < vertices; i++) {
            adjacencyList[i] = new DoublyLinkedList<>();
        }
    }

    public void addEdge(int source, int destination) {
        if (source < vertices && destination < vertices) {
            this.adjacencyList[source].insertAtEnd(destination);

        //for undirected graph uncomment the line below
        // this.adjacencyList[destination].insertAtEnd(source);
        }
    }

    public void printGraph()
    {
        System.out.println(">>Adjacency List of Directed Graph<<");
        for (int i = 0; i < vertices; i++)
        {
            if(adjacencyList[i]!=null){
                System.out.print("|" + i + "| => ");

                DoublyLinkedList<Integer>.Node temp = adjacencyList[i].getHeadNode();
                while (temp != null)
                {
                    System.out.print("[" + temp.data + "] -> ");
                    temp = temp.nextNode;
                }
                System.out.println("null");
            }
            else{

                System.out.println("|" + i + "| => "+ "null");
            }
        }
    }

    public static void main( String args[] ) {
        Graph g= new Graph(4);
        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 3);
        g.addEdge(2, 3);

        g.printGraph();
        /**
            >>Adjacency List of Directed Graph<<
            |0| => [1] -> [2] -> null
            |1| => [3] -> null
            |2| => [3] -> null
            |3| => null
        */
    }
}

